truffle/src/main/java/org/jruby/truffle/core/VMPrimitiveNodes.java

```diff
@@ -263,6 +273,66 @@ public boolean vmObjectRespondTo(VirtualFrame frame, Object object, Object name,
 
     }
 
+    @Primitive(name = "vm_object_respond_to_no_built_in", needsSelf = false)
+    public static abstract class VMObjectRespondToNoBuiltInPrimitiveNode extends PrimitiveArrayArgumentsNode {
+
+        @Child NameToJavaStringNode nameToJavaStringNode;
+
+        public VMObjectRespondToNoBuiltInPrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            nameToJavaStringNode = NameToJavaStringNode.create();
+        }
+
+        @Specialization
+        public boolean vmObjectRespondToNoBuiltIn(VirtualFrame frame, Object object, Object name, boolean includePrivate) {
+            // Used in place for this pattern:
+            // method_entry_get(klass, mid, &defined_class);
+            // if (!me || METHOD_ENTRY_BASIC(me))
+
+            final String nameString = nameToJavaStringNode.executeToJavaString(frame, name);
+            final InternalMethod method = ModuleOperations.lookupMethod(coreLibrary().getMetaClass(object), nameString);
+
+            // If no method was found, use #method_missing
+            if (method == null) {
+                return false;
+            }
+
+            // Check for methods that are explicitly undefined
+            if (method.isUndefined()) {
+                return false;
+            }
+
+            if(method.isBuiltIn()){
+                return false;
+            }
+
+            return true;
+        }
+
+    }
```

Comment:
I think this would be easier to follow if it returned a Ruby Method or nil.
And maybe a separate primitive to check if a Method is built-in or not.